<!DOCTYPE html>
<html lang="en">
	<head>
		<meta charset="UTF-8" />
		<meta http-equiv="X-UA-Compatible" content="IE=edge" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
		<title>Document</title>
	</head>
	<body>
		<script>
			// 先定义三个常量表示状态
			const PENDING = 'pending'
			const FULFILLED = 'fulfilled'
			const REJECTED = 'rejected'

			// 新建 MyPromise 类
			class MyPromise {
				constructor(executor) {
					// execute 是一个执行器，进入会立即执行
					try {
						executor(this.resolve, this.reject)
					} catch (error) {
						resolve(error)
					}
				}

				status = PENDING

				// 成功之后的值
				value = null
				// 失败之后的原因
				reason = null

				// 存储成功回调
				onFulfilledCallback = []
				// 存储失败回调
				onRejectedCallback = []

				// 更新成功后的状态
				resolve = (value) => {
					if (this.status === PENDING) {
						this.status = FULFILLED
						this.value = value
						while (this.onFulfilledCallback.length)
							this.onFulfilledCallback.shift()(value)
					}
				}

				// 更新失败后的状态
				reject = (reason) => {
					if (this.status === PENDING) {
						this.status = REJECTED
						this.reason = reason
						while (this.onRejectedCallback.length)
							this.onRejectedCallback.shift()(reason)
					}
				}

				static resolve(parmeter) {
					if (parmeter instanceof MyPromise) {
						return parmeter
					}

					return new MyPromise((resolve) => {
						resolve(parmeter)
					})
				}
				static rejected(parmeter) {
					return new MyPromise((resolve, rejected) => {
						rejected(parmeter)
					})
				}
				then(onFulfilled, onRejected) {
					onFulfilled =
						typeof onFulfilled === 'function' ? onFulfilled : (value) => value
					onRejected =
						typeof onRejected === 'function'
							? onRejected
							: (reason) => {
									throw reason
							  }
					const promise2 = new MyPromise((resolve, reject) => {
						if (this.status === FULFILLED) {
							// onFulfilled(this.value)

							// ==== 新增 ====
							// 创建一个微任务等待 promise2 完成初始化
							queueMicrotask(() => {
								// ==== 新增 ====
								try {
									// 获取成功回调函数的执行结果
									const x = onFulfilled(this.value)
									// 传入 resolvePromise 集中处理
									resolvePromise(promise2, x, resolve, reject)
								} catch (error) {
									reject(error)
								}
							})
						} else if (this.status === REJECTED) {
							// 调用失败回调，并且把原因返回
							// onRejected(this.reason)
							queueMicrotask(() => {
								try {
									const x = onRejected(this.reason)
									resolvePromise(promise2, x, resolve, reject)
								} catch (error) {
									reject(error)
								}
							})
						} else if (this.status === PENDING) {
							// 等待
							// 因为不知道后面状态的变化情况，所以将成功回调和失败回调存储起来
							// 等到执行成功失败函数的时候再传递
							this.onFulfilledCallbacks.push(() => {
								// ==== 新增 ====
								queueMicrotask(() => {
									try {
										// 获取成功回调函数的执行结果
										const x = onFulfilled(this.value)
										// 传入 resolvePromise 集中处理
										resolvePromise(promise2, x, resolve, reject)
									} catch (error) {
										reject(error)
									}
								})
							})
							this.onRejectedCallbacks.push(() => {
								// ==== 新增 ====
								queueMicrotask(() => {
									try {
										// 调用失败回调，并且把原因返回
										const x = onRejected(this.reason)
										// 传入 resolvePromise 集中处理
										resolvePromise(promise2, x, resolve, reject)
									} catch (error) {
										reject(error)
									}
								})
							})
						}
					})
					return promise2
				}
			}

			function resolvePromise(promise2, x, resolve, rejected) {
				if (promise2 === x) {
					return rejected(
						new TypeError('Chaining cycle detected for promise #<Promise>')
					)
				}
				if (x instanceof MyPromise) {
					x.then(resolve, rejected)
				} else {
					resolve(x)
				}
			}
		</script>
	</body>
</html>
